Source/Core/VideoCommon/GXPipelineTypes.h

```diff
@@ -35,24 +35,6 @@ struct GXPipelineUid
   // and this map lookup can happen every draw call. However, as using memcmp() will also compare
   // any padding bytes, we have to ensure these are zeroed out.
   GXPipelineUid() { std::memset(static_cast<void*>(this), 0, sizeof(*this)); }
-#ifdef _MSC_VER
```

Comment:
Is this safe? The padding bytes are guaranteed to be zero when a GXPipelineUid is constructed, but can they end up being different from zero at some point? If so, the default copy constructor and copy assignment operator are not guaranteed to copy the padding bytes.

Reply:
Good point; I can only see it being a problem if the memset will not run before some other constructor, is that the case? If so, we could initialize members at declaration instead and pack the structure, or other things.

Reply:
The memset runs after every member is constructed, but I was wondering if something could change the padding bytes *after* the memset. For example:

```cpp
GXPipelineUid uid;
// default ctor sets everything (including padding bytes) to 0

// initialise uid
...
// can anything we do here (e.g. assigning to members) affect the padding bytes?

GXPipelineUid copy = uid;
// are we guaranteed to have copy == uid?
```

Another potential problem I see is that unlike with unions, the default copy ctor for a class/struct performs a "memberwise copy" (as opposed to copying the object representation), so I don't think it's guaranteed that the padding bytes will be copied.

In practice, GCC and Clang [seem to generate copy constructors that copy the entire struct including padding bytes](https://godbolt.org/z/aPK59qh7s), but I'm not sure if this is something we should rely on. It might be safer to keep the memcpy to force the object representation to be copied.

Reply:
Sorry, I know what you meant (re: padding modification after construction). My response was to the point that the padding can only be changed if an object has been constructed with indeterminate padding contents, and that object gets fully copied over existing (zeroized) padding. However, such object with indeterminate padding would be the "first order" problem itself, not the copy to the secondary object.

tbh I think most of the code here is a bit gross and reaching for some optimization which doesn't make sense. For instance the objects are not aligned nor packed, undefined behavior is being relied on and I'm sure a simple memberwise comparison could be faster than memcmp in the general case, anyway. As well, in-class member initialization would be nicer. In any case, my goal is only to get rid of the code special cased for msvc.

Reply:
> My response was to the point that the padding can only be changed if an object has been constructed with indeterminate padding contents, and that object gets fully copied over existing (zeroized) padding. However, such object with indeterminate padding would be the "first order" problem itself, not the copy to the secondary object.

Ah, removing the memcpy sounds fine then.

> tbh I think most of the code here is a bit gross and reaching for some optimization which doesn't make sense. For instance the objects are not aligned nor packed, undefined behavior is being relied on and I'm sure a simple memberwise comparison could be faster than memcmp in the general case, anyway.

Yeah, definitely. Adding the memset and memcmp actually pessimized GCC's codegen for constructing and operator== (though neither Clang nor GCC actually emit calls to library functions). I wonder if this was different when this code was added a few years ago.

> In any case, my goal is only to get rid of the code special cased for msvc.

Only the warning stuff is behind the MSVC ifdef -- it looks like the intent was to use memcpy even for non-MSVC compilers.

Reply:
OK, I guess the PR seems ready then..?